### ISSUE TITLE
NuGet.Server 3.0.0 having problems with delist

### ISSUE THREAD

**Issue by TiltonJH** [NuGet/NuGetGallery#5066]
Last week we switched from NuGet.Server 2.11.3 to the new 3.0.0. Since then all our hidden (delisted) packages are showing up again. (FYI: Our delisted packages are all prerelease)

We use VS 2015 with NuGet - VS 2015 VSIX - latest v3.6.0 and the current VS 2017. Both show delisted packages.

How ever the NuGet.exe (4.4.1.4656) does not show any delisted packages. Not even when we force it via "-IncludeDelisted".

The delisting itself does seam to work. The *.nupkg-file gets the hidden attribute.

Furthermore, now that I have taken a look at the git repo for the NuGet server, there is no release for 3.0.0 it seems to be still in betha, while nuget.org shows that it is not a prerelease?

**Comment by joelverhagen** [comment]
Hey @TiltonJH, thanks for reaching out about this.

I have added the 3.0.0 release to GitHub.
https://github.com/NuGet/NuGet.Server/releases/tag/3.0.0

Concerning the `-EnablePrerelease`, I am pretty sure that flag has some confusing caveats. I would recommend asking at https://github.com/NuGet/Home/issues about that.

I tried the latest NuGet.Server locally (3.0.0) and could not reproduce the issue you described, neither with pre-release or stable packages.

I followed the following steps:

1. Open Visual Studio 2017.
1. Create a new empty .NET Framework 4.6 web project.
1. Install NuGet.Server 3.0.0
1. Edit the web.config to add an `apiKey` app setting and set `enableDelisting` to true.
1. F5 to launch to the server.
1. Use nuget.exe `push` and `delete` to push and unlist two packages.
1. Open another VS and add the local server as a package source.
1. Manage NuGet packages in a new project, select the local server, and go to the Browse tab.
1. Run `nuget.exe list` against the local server.

Works okay for me. Could you answer/do the following?

- Try the minimal repro steps above and see if you run into an issue.
- Are only newly pushed packages a problem (i.e. packages pushed after updating to 3.0.0)? Or all?
- What does `/nuget/Packages()` show the packages in question as listed or unlisted? This is an XML document but relatively easy to read as a human.

**Comment by TiltonJH** [comment]
My Steps:

1. Open Visual Studio 2015.
2. Create a new empty .NET Framework 4.6 web project.
3. Install NuGet.Server 3.0.0
4. Edit the web.config to add an apiKey app setting and set enableDelisting to true and change the packagesPath to a local path like "N:\NuGetPackageSources" (BTW: FileSystem: ReFS)
5. Published a Release Build to my local file system.
6. Stopped the IIS app pool on the Win2012R2
7. Moved published files over to the server
8. Started the IIS app pool 
9. Cleared the cash
10. Viewed `/nuget/Packages()` -> shows <d:Listed m:type="Edm.Boolean">false</d:Listed> for the unlisted packages
11. Viewed the `http://<ourserver.name>/<SomeSubDir>/NuGetInQu/nuget` in VS2015 and VS2017 -> both show unlisted packages in `Browse`
12. Viewed `nuget.exe list -allversions` -> unlisted packages are not shown
13. Viewed `nuget.exe list -allversions -Includedelisted` -> unlisted packages are not shown

I tried the same thing with NuGet.Server 2.11.3 and everything (besides the `nuget.exe list -allversions -Includedelisted`) workes as it should.

All packages have that problem.

A response example from NuGet.Server 3.0.0 for `http://ourserver.name/SomeSubDir/nuget/Packages(Id='NuGet.Server',Version='2.11.3')`
[3.0.0.xml.txt](https://github.com/NuGet/NuGetGallery/files/1511409/3.0.0.xml.txt)

and from 2.11.3 for `http://ourserver.name/SomeSubDirTest/nuget/Packages(Id='NuGet.Server',Version='2.11.3')`:
[2.11.3.xml.txt](https://github.com/NuGet/NuGetGallery/files/1511408/2.11.3.xml.txt)

**Comment by joelverhagen** [comment]
Thanks for the additional information. The fact that nuget.exe and Visual Studio behave differently is very odd. Could you provide [Fiddler](https://www.telerik.com/fiddler) traces of interacting with both the 2.11.3 and 3.0.0 server using both nuget.exe and Visual Studio?

Also, could you provide the response body when hitting the following URL on both servers?
```
/nuget/Search()?$filter=IsAbsoluteLatestVersion&searchTerm=''&targetFramework='net46'&includePrerelease=true&$skip=0&$top=26
```

This is basically the initial search done by VS, Browse tab.

Thanks for the help so far!

**Comment by TiltonJH** [comment]
**a) used Url**: `/nuget/Search()?$filter=IsAbsoluteLatestVersion&searchTerm='nuget'&targetFramework='net46'&includePrerelease=true&$skip=0&$top=26` (Because we have over about 3000 packages in that one server, I have adjusted the url to specify a `searchTerm`.)

**b) used NuGet.exe command:** `nuget list nuget -Source http://ourserver.name/SomeSubDirTest/nuget -allversion`

**c) used NuGet.exe command:** `nuget list nuget -Source http://ourserver.name/SomeSubDirTest/nuget`
(== `/nuget/Search()?$filter=IsLatestVersion&$orderby=Id&searchTerm='nuget'&targetFramework=''&includePrerelease=false&$skip=0&$top=30&semVerLevel=2.0.0'`)

**d) using VS20115**

The test uses the package `NuGet.Server` of which we have version `2.11.3` and `3.0.0`.
(BTW: I have noticed that with or without perelease doesn't matter, both show same behavior.)

### Only delisted version **`2.11.3`** on the `3.0.0` server.
**a)** [3.0.0.with.2.11.3.delisted.xml.txt](https://github.com/NuGet/NuGetGallery/files/1514436/3.0.0.with.2.11.3.delisted.xml.txt)
--> **correct**

**b)**
`NuGet.Core 2.14.0`
`NuGet.Server 3.0.0`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **correct**

**c)**
`NuGet.Core 2.14.0`
`NuGet.Server 3.0.0`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **correct**

**d)** shows the delisted Version `2.11.3`: --> **failed**
![vs2015 with 3 0 0](https://user-images.githubusercontent.com/5372425/33384779-41b00672-d527-11e7-89dd-e55f1d482268.png)



### Only delisted version **`3.0.0`** on the `3.0.0` server.
**a)** [3.0.0.with.3.0.0.delisted.xml.txt](https://github.com/NuGet/NuGetGallery/files/1514446/3.0.0.with.3.0.0.delisted.xml.txt)
--> **failed** -> now doesn't show any `NuGet.Server` packages; should at least show version `2.11.3`

**b)**
`NuGet.Core 2.14.0`
`NuGet.Server 2.11.3`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **correct**

**c)**
`NuGet.Core 2.14.0`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **failed** (may be? Since `NuGet.Server` `2.11.3` is the highest listed package it should show up.)

**d)** doesn't show any `NuGet.Server` packages while it should show at least version `2.11.3`.

### Only delisted version **`2.11.3`** on the `2.11.3` server.
**a)** [2.11.3.with.2.11.3.delisted.xml.txt](https://github.com/NuGet/NuGetGallery/files/1514448/2.11.3.with.2.11.3.delisted.xml.txt)
--> **correct**

**b)**
`NuGet.Core 2.14.0`
`NuGet.Server 3.0.0`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **correct**

**c)**
`NuGet.Core 2.14.0`
`NuGet.Server 3.0.0`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **correct**

**d)** --> **correct**; does not show the delisted >Version `2.11.3`:
![vs2015 with 2 11 3](https://user-images.githubusercontent.com/5372425/33384838-64dc5f10-d527-11e7-90bd-93b52e891c4c.png)


### Only delisted version **`3.0.0`** on the `2.11.3` server.
**a)** [2.11.3.with.3.0.0.delisted.xml.txt](https://github.com/NuGet/NuGetGallery/files/1514450/2.11.3.with.3.0.0.delisted.xml.txt)
--> **failed** -> now doesn't show any `NuGet.Server` packages; should at least show version `2.11.3`

**b)**
`NuGet.Core 2.14.0`
`NuGet.Server 2.11.3`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **correct**

**c)**
`NuGet.Core 2.14.0`
`NuGet.Server.Core 3.0.0`
`NuGet.Server.V2 3.0.0`
--> **failed** (again may be? Since `NuGet.Server` `2.11.3` is the highest listed package it should show up.)

**d)** doesn't show any `NuGet.Server` packages while it should show at least version `2.11.3`.

### Thoughts
I would have expected that the nuget server would do exactly the same thing when a package gets delisted or if the package gets deleted. With the exception that a delisted package can be restored. But that does not seam to be the case. So is this a flaw or by design?

FYI: Our problem orginaly showed up like that:
We had a stable release `1.0.6`. We made a new prerelease `1.1.0-Dev` and push it to the server. On second though we decides not to use the prerelease and delisted it. Now our projects using the prerelease would still be able to restore and build, but the usage of the faulty prerelease would not spread any further. After we downgraded all projects using the prerelease, we forgot it ever had existed. Over time other prerelease came along but with versions like `1.0.7-Dev`. Still everything was fine, because the faulty delisted prerelease did not show up as an update.
The problem began after switching to the 3.0.0 NuGet.Server when browse and update now showed the faulty actually delisted prerelease.

**Comment by joelverhagen** [comment]
I noticed your XML files (e.g. 3.0.0.with.2.11.3.delisted.xml.txt) did not include NuGet.Core. Were you actually searching for `nuget.server` instead of `nuget`? Aside from this, I fully understand the situation now. It was thinking that a unlisted version was showing up in the search results panel (the left panel in the VS Browse tab). I did not pick up on that fact that you were actually talking about the package details panel (the right panel). Your screenshots were very helpful.

Concerning `2.11.3` disappearing when `3.0.0` is unlisted, this is a known bug: https://github.com/NuGet/NuGetGallery/issues/3434. 

Concerning unlisted versions showing up in the dropdown in VS, this is a know bug: https://github.com/NuGet/NuGetGallery/issues/4020.

Once these are fixed, I think all of your problems will be resolved.

> I would have expected that the nuget server would do exactly the same thing when a package gets delisted or if the package gets deleted. With the exception that a delisted package can be restored. But that does not seam to be the case. So is this a flaw or by design?

The only endpoint that `listed` effects is the `Search()` endpoint. `FindPackagesById()`, `Packages()`, and `Package(Id='{ID}',Version='{VERSION}')` all return unlisted packages as these are used for NuGet restore and operations where the version is known.

> Are there any other fiddler like tools I could use? I can not use fiddler on company pc.

You could theoretically implement a simple proxy in C# and redirect NuGet traffic through it ([docs](https://docs.microsoft.com/en-us/nuget/schema/nuget-config-file#config-section)) but this would be a hassle. If you want to debug HTTP requests on the command line (`nuget.exe list`) you can use the `-Verbosity detailed` switch to see the URLs being queries. This is not perfect but better than nothing.

To summarize, NuGet.Server 3.0.0 needs a fix so unlisted packages are properly filtered out from the VS UI. Also, when a package is unlisted, the `IsLatest` properties should be re-evaluated.

**Comment by joelverhagen** [comment]
Duplicate of https://github.com/NuGet/NuGetGallery/issues/3434 and https://github.com/NuGet/NuGetGallery/issues/4020.

**Comment by joelverhagen** [comment]
Hey @TiltonJH, I've fixed the bugs you ran into:
https://www.nuget.org/packages/NuGet.Server/2.14.2
https://www.nuget.org/packages/NuGet.Server/3.0.1

Let me know if you run into any other issues.

**Comment by TiltonJH** [comment]
Hey Joel,

 

version 3.0.1 fixes all my current NuGet.Server issues. :-)

 

Thx

 

Gesendet: Freitag, 01. Dezember 2017 um 23:27 Uhr
Von: "Joel Verhagen" <notifications@github.com>
An: NuGet/NuGetGallery <NuGetGallery@noreply.github.com>
Cc: TiltonJH <Vasco.Wagner@gmx.net>, Mention <mention@noreply.github.com>
Betreff: Re: [NuGet/NuGetGallery] NuGet.Server 3.0.0 having problems with delist (#5066)


Hey @TiltonJH, I've fixed the bugs you ran into:
https://www.nuget.org/packages/NuGet.Server/2.14.2
https://www.nuget.org/packages/NuGet.Server/3.0.1

Let me know if you run into any other issues.
